python/pyarrow/_dataset.pyx

```diff
@@ -806,173 +884,88 @@ cdef class ScanTask:
             CRecordBatchIterator iterator
             shared_ptr[CRecordBatch] record_batch
 
-        iterator = move(GetResultValue(move(self.task.Execute())))
-
-        while True:
-            record_batch = GetResultValue(iterator.Next())
-            if record_batch.get() == nullptr:
-                raise StopIteration()
-            else:
-                yield pyarrow_wrap_batch(record_batch)
-
-
-cdef class InMemoryScanTask(ScanTask):
-    """A trivial ScanTask that yields the RecordBatch of an array."""
-
-    cdef:
-        CInMemoryScanTask* in_memory_task
+        with nogil:
+            iterator = move(GetResultValue(move(self.task.Execute())))
 
-    cdef init(self, shared_ptr[CScanTask]& sp):
-        ScanTask.init(self, sp)
-        self.in_memory_task = <CInMemoryScanTask*> sp.get()
+            while True:
+                record_batch = GetResultValue(iterator.Next())
+                if record_batch.get() == nullptr:
+                    raise StopIteration()
+                else:
+                    with gil:
+                        yield pyarrow_wrap_batch(record_batch)
 
 
-cdef class ScannerBuilder:
-    """Factory class to construct a Scanner.
+cdef class Scanner:
+    """A materialized scan operation with context and options bound.
 
-    It is used to pass information, notably a potential filter expression and a
-    subset of columns to materialize.
+    A scanner is the class that glues the scan tasks, data fragments and data
+    sources together.
 
     Parameters
     ----------
-    dataset : Dataset
-        The dataset to scan.
```

Comment:
it still takes the dataset as first argument?

Reply:
Yep, nice catch! Added the dataset parameter back to the docstring.